### ISSUE TITLE
dead_code lint does not trigger for tuple structs with unread fields

### ISSUE THREAD

**Issue by shepmaster** [rust-lang/rust#92790]
This code:

```rust
fn main() {
    let _ = Example { u: 0 };
}

struct Example {
    u: u8,
}
```

Produces this warning:

```
warning: field is never read: `u`
 --> src/main.rs:6:5
  |
6 |     u: u8,
  |     ^^^^^
  |
```

However, switching to a tuple struct:

```rust
fn main() {
    let _ = Example(0);
}

struct Example(u8);
```

Does not report a warning.

<details>
<summary>Original report misleadingly focusing on <code>Debug</code></summary>

```rust
fn main() {
    dbg!(Example { a: 0 });
}

#[derive(Debug)]
struct Example {
    a: u8,
}
```

produces 

```
warning: field is never read: `a`
 --> src/main.rs:7:5
  |
7 |     a: u8,
  |     ^^^^^
  |
  = note: `#[warn(dead_code)]` on by default
```

However, switching to a tuple struct:

```rust
fn main() {
    dbg!(Example(0));
}

#[derive(Debug)]
struct Example(u8);
```

does not produce the warning.

/cc @FabianWolff
/cc #85200; #84647; #88900

</details>

### Meta
Rustc 1.57.0 and 1.60.0-nightly (2022-01-10 89b9f7b284aacc5f8613)

**Comment by FabianWolff** [comment]
This has nothing to do with the `#[derive(Debug)]`:
```rust
fn main() {
    let _ = Example { u: 0 };
}

struct Example {
    u: u8,
}
```
```
warning: field is never read: `u`
 --> src/main.rs:6:5
  |
6 |     u: u8,
  |     ^^^^^
  |
```
but
```
fn main() {
    let _ = Example(0);
}

struct Example(u8);
```
gives no warning.

**Comment by shepmaster** [comment]
@FabianWolff oh dear. Thanks! I've updated the title and original post; sorry for the false ping.

**Comment by camelid** [comment]
Not a regression AFAIK, but seems like worth prioritizing.

**Comment by shepmaster** [comment]
> Not a regression AFAIK

I agree. I did a quick test across Rust 1.{0,10,20,30,40,50}.0 and none of them reported the lint.

**Comment by camelid** [comment]
Thanks for checking!

**Comment by FabianWolff** [comment]
This behavior does indeed date back to the very beginning of the detection of dead struct fields (commit https://github.com/rust-lang/rust/commit/0271224bdae26260ab498f47323997f9edb5879e in 2014). The culprit is line 621 in `dead.rs`:
https://github.com/rust-lang/rust/blob/72e74d7b9cf1a7901650227e74650f1fcc797600/compiler/rustc_passes/src/dead.rs#L618-L625

Commenting out this line with the above tuple struct example gives:
```
warning: field is never read: `0`
 --> s2.rs:6:16
  |
6 | struct Example(u8);
  |                ^^
  |
```
I can open a PR for this, or is there a reason anyone here can think of why dead tuple struct fields _shouldn't_ get a warning?

@rustbot claim

**Comment by camelid** [comment]
I can't think of one, so opening a PR seems good.

**Comment by danielhenrymantilla** [comment]
FWIW, I find the example with `dbg!` to be surprising: `dbg!` is debug-printing a struct, and thus reading its fields

**Comment by shepmaster** [comment]
@danielhenrymantilla see #85200; #84647; #88900

**Comment by apiraino** [comment]
Assigning priority as discussed in the [Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2392790.20dead_code.20lint.20does.20not.20trigger.20for.20tuple.20structs.20wi.E2.80.A6) of the Prioritization Working Group.

@rustbot label -I-prioritize +P-medium

**Comment by rcls** [comment]
> is there a reason anyone here can think of why dead tuple struct fields _shouldn't_ get a warning?

One thought is that deleting a dead tuple field may be non-trivial, if it is not the last field. E.g., imagine if someone got:
```
struct DeadLive(u8,u8);
...
warning: field is never read: `0`
```
You can't simply delete the dead field. You have to also renumber all the uses of `.1`, which might be a non-trivial and error prone task.  So the cost v. benefit tradeoff of having a warning is different to the named-field case.

**Comment by shepmaster** [comment]
Thank you @FabianWolff !
